### DISCUSSION TITLE
Using getServerSession in Express middleware and Prisma adapter

### DISCUSSION THREAD

**Discussion by nathanbirrell** [nextauthjs/next-auth#6849] [Help]
### Question 💬

I have a fairly straightforward NextAuth implementation running on a node Express container. How would I manage to get `getServerSession` functional within my Express middleware? Sessions are stored in Postgres via the Prisma adapter.

The middleware will be used to apply some minor basic role-based access controls on our GraphQL API.

### How to reproduce ☕️

Some details on my setup:

`"next-auth": "^4.19.2",` + `"@next-auth/prisma-adapter": "^1.0.5",`

My Express app entrypoint:

```ts
app.use(NextAuthMiddleware())

app.use(AuthorizationMiddleware())
```

and my auth.ts:

```ts
const OPTIONS: NextAuthOptions = {
  secret: process.env.NEXTAUTH_SECRET,
  adapter: PrismaAdapter(db),
  session: {
    strategy: 'database'
  },
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID ?? '',
      clientSecret: process.env.GOOGLE_CLIENT_SECRET ?? ''
    }),
  ]
}

/**
 * Should match the following paths:
 * /api/auth/signin
 * /api/auth/signin/:provider
 * /api/auth/callback/:provider
 * /api/auth/signout
 * /api/auth/session
 * /api/auth/csrf
 * /api/auth/providers
 * /api/auth/_log
 *
 * See: https://next-auth.js.org/getting-started/rest-api
 */
const authActions = /^\/api\/auth\/(session|signin\/?\w*|signout|csrf|providers|callback\/\w+|_log)$/

const router = Router()

/**
 * Compatibility layer for NextAuth + Express
 *    - Requests come in to Express, here we mangle them into something NextAuth expects
 *    - NextAuth returns actual HTML pages for the various auth screens as well as the typical REST endpoints
 */
export function NextAuthMiddleware(): Router {
  return router
    .use(urlencoded({ extended: false }))
    .use(json())
    .use(cookieParser())
    .all(authActions, (req: IncomingMessage, res: ServerResponse, next) => {
      if (req.method !== 'POST' && req.method !== 'GET') {
        next()
        return
      }
      // @ts-expect-error
      req.query.nextauth = req.path.split('/').slice(3)
      // @ts-expect-error
      NextAuth(req, res, OPTIONS)
    })
}

export const AuthorizationMiddleware = () => {
  return router
    .use(urlencoded({ extended: false }))
    .use(json())
    .use(cookieParser())
    .all('/graphql', async (request: IncomingMessage, response: ServerResponse, next) => {

      const session = await getServerSession(request, response, OPTIONS)

      // 🚫 Session always = null

      next()
    })
}
```

## Expected behavior

`const session = await getServerSession(request, response, OPTIONS)` should return a valid session if there is one.

## Actual behavior

Session is always `null`.

`{ session: null }`

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by balazsorban44** [comment]
`getServerSession` is designed for Next.js requests and responses, not Express.js'.

https://github.com/nextauthjs/next-auth/blob/0a689b4f4ee9394d20ec2652c197977166727ba2/packages/next-auth/src/next/index.ts#L149

Maybe cookies are not parsed correctly. Or somehow the session cookie is not forwarded to the graphql endpoint.

Although the rest looks good.

Another thing, `NextAuth` returns a promise, maybe you need to await it properly. 🤔
